### ISSUE TITLE
Network Finetuning GPU memory usage

### ISSUE THREAD

**Issue by lucabergamini** [pytorch/pytorch#2672]
I'm a bit puzzled about the gpu memory usage when finetuning a network. In particular if I use a VGG-19 model with every conv layer frozen, **where should I see the difference in GPU consumption**? In forward or in backward? I'm running two models, one with frozen layers and one without but, as far as nvidia-smi tells me, if I place a breakpoint **after** the forward call , the two models have allocated exactly the same GPU memory. 
I will provide reproducible code for both cases in the afternoon, but to achieve the finetuning I just set 
```requires_grad = False``` for every layer I'm not interested in, and excluded them from the optimizer parameters.

**Comment by Stonesjtu** [comment]
Some code is not optimized for the inference at all. Actually I would like to think the `requires_grad` as an attribute for `Optimizer` rather than an attribute for forward and backward in **current** version(0.2) of `pytorch`, but sometimes it helps for error detecting, e.g. the `index` for `index_select` can't have `requires_grad=True`.
 Let's take a look at the sample code for `cumprod`, the `save_for_backward` function saves the operands regardless of `requires_grad`.
https://github.com/pytorch/pytorch/blob/459cc5a346e25200eca1b921729d10363fac2b1b/torch/autograd/_functions/tensor.py#L735-L742

I don't exactly know about how convolution is performed by low level matrix operation, but this should be one problem.

**Comment by lucabergamini** [comment]
My bad, I was setting ```requires_grad=True``` on the input variable, thus forcing the network to backprop to it.
 **However setting it to False change the accuracy results**...is it expected?

**Comment by soumith** [comment]
setting it to false shouldn't change accuracy results. Make sure you set the model to `model.eval()` maybe?

**Comment by apaszke** [comment]
If you have a script that can reproduce the accuracy problems please open a new issue. I'll close this.
